File: android/app/src/main/java/com/budgetmanager/domain/interactor/ExpenseInteractor.kt
package com.budgetmanager.domain.interactor

import com.budgetmanager.domain.entity.DomainError
import com.budgetmanager.domain.entity.Expense
import com.budgetmanager.domain.repository.BudgetRepository
import com.budgetmanager.domain.repository.ExpenseRepository
import java.time.LocalDate

class ExpenseInteractor(
    private val expenseRepository: ExpenseRepository,
    private val budgetRepository: BudgetRepository
) {
    suspend fun getExpenses(): List<Expense> {
        return expenseRepository.getAll()
    }

    suspend fun getExpense(id: Long): Expense {
        return expenseRepository.getById(id)
    }

    suspend fun getExpensesByBudgetId(budgetId: Long): List<Expense> {
        return expenseRepository.getByBudgetId(budgetId)
    }

    suspend fun getExpensesByDateRange(startDate: LocalDate, endDate: LocalDate): List<Expense> {
        return expenseRepository.getByDateRange(startDate, endDate)
    }

    suspend fun createExpense(
        budgetId: Long,
        amount: Double,
        description: String,
        date: LocalDate = LocalDate.now()
    ): Expense {
        if (amount <= 0) {
            throw DomainError.InvalidInput("Amount must be greater than 0")
        }
        if (description.isBlank()) {
            throw DomainError.InvalidInput("Description cannot be empty")
        }

        val budget = budgetRepository.getById(budgetId)
        // Skip insufficient funds check for Uncategorized budget (allocatedAmount = 0)
        // This allows expenses to be added without pre-allocation
        if (budget.allocatedAmount > 0 && budget.remainingAmount < amount) {
            throw DomainError.InsufficientFunds
        }

        val expense = Expense(
            budgetId = budgetId,
            amount = amount,
            description = description,
            date = date
        )

        val createdExpense = expenseRepository.add(expense)
        budgetRepository.updateSpentAmount(budgetId, amount)

        return createdExpense
    }

    suspend fun updateExpense(
        id: Long,
        budgetId: Long? = null,
        amount: Double? = null,
        description: String? = null,
        date: LocalDate? = null
    ): Expense {
        val expense = expenseRepository.getById(id)
        val oldAmount = expense.amount
        val oldBudgetId = expense.budgetId

        val updatedExpense = expense.copy(
            budgetId = budgetId ?: expense.budgetId,
            amount = amount ?: expense.amount,
            description = description ?: expense.description,
            date = date ?: expense.date
        )

        if (updatedExpense.budgetId != oldBudgetId) {
            val newBudget = budgetRepository.getById(updatedExpense.budgetId)
            if (newBudget.remainingAmount < updatedExpense.amount) {
                throw DomainError.InsufficientFunds
            }

            budgetRepository.updateSpentAmount(oldBudgetId, -oldAmount)
            budgetRepository.updateSpentAmount(updatedExpense.budgetId, updatedExpense.amount)
        } else if (updatedExpense.amount != oldAmount) {
            val diff = updatedExpense.amount - oldAmount
            val budget = budgetRepository.getById(updatedExpense.budgetId)

            if (diff > 0 && budget.remainingAmount < diff) {
                throw DomainError.InsufficientFunds
            }

            budgetRepository.updateSpentAmount(updatedExpense.budgetId, diff)
        }

        expenseRepository.update(updatedExpense)
        return updatedExpense
    }

    suspend fun deleteExpense(id: Long) {
        val expense = expenseRepository.getById(id)
        budgetRepository.updateSpentAmount(expense.budgetId, -expense.amount)
        expenseRepository.delete(id)
    }
}

File: iOS/Sources/Domain/Interactors/ExpenseInteractor.swift
import Foundation

class ExpenseInteractor {
    private let expenseRepository: ExpenseRepositoryProtocol
    private let budgetRepository: BudgetRepositoryProtocol

    init(expenseRepository: ExpenseRepositoryProtocol, budgetRepository: BudgetRepositoryProtocol) {
        self.expenseRepository = expenseRepository
        self.budgetRepository = budgetRepository
    }

    func getExpenses() async throws -> [Expense] {
        return try await expenseRepository.fetchAll()
    }

    func getExpense(id: Int64) async throws -> Expense {
        return try await expenseRepository.fetch(id: id)
    }

    func getExpenses(forBudgetId budgetId: Int64) async throws -> [Expense] {
        return try await expenseRepository.fetchByBudgetId(budgetId)
    }

    func getExpenses(from startDate: Date, to endDate: Date) async throws -> [Expense] {
        return try await expenseRepository.fetchByDateRange(start: startDate, end: endDate)
    }

    func createExpense(
        budgetId: Int64,
        amount: Double,
        description: String,
        date: Date = Date()
    ) async throws -> Expense {
        guard amount > 0 else {
            throw DomainError.invalidInput("Amount must be greater than 0")
        }
        guard !description.isEmpty else {
            throw DomainError.invalidInput("Description cannot be empty")
        }

        let budget = try await budgetRepository.fetch(id: budgetId)
        guard budget.remainingAmount >= amount else {
            throw DomainError.insufficientFunds
        }

        let expense = Expense(
            budgetId: budgetId,
            amount: amount,
            description: description,
            date: date
        )

        let createdExpense = try await expenseRepository.add(expense)
        try await budgetRepository.updateSpentAmount(id: budgetId, amount: amount)

        return createdExpense
    }

    func updateExpense(
        id: Int64,
        budgetId: Int64?,
        amount: Double?,
        description: String?,
        date: Date?
    ) async throws -> Expense {
        var expense = try await expenseRepository.fetch(id: id)
        let oldAmount = expense.amount
        let oldBudgetId = expense.budgetId

        if let description = description {
            expense.description = description
        }
        if let date = date {
            expense.date = date
        }
        if let amount = amount {
            expense.amount = amount
        }
        if let budgetId = budgetId {
            expense.budgetId = budgetId
        }

        if expense.budgetId != oldBudgetId {
            let newBudget = try await budgetRepository.fetch(id: expense.budgetId)
            guard newBudget.remainingAmount >= expense.amount else {
                throw DomainError.insufficientFunds
            }

            try await budgetRepository.updateSpentAmount(id: oldBudgetId, amount: -oldAmount)
            try await budgetRepository.updateSpentAmount(id: expense.budgetId, amount: expense.amount)
        } else if expense.amount != oldAmount {
            let diff = expense.amount - oldAmount
            let budget = try await budgetRepository.fetch(id: expense.budgetId)

            if diff > 0 && budget.remainingAmount < diff {
                throw DomainError.insufficientFunds
            }

            try await budgetRepository.updateSpentAmount(id: expense.budgetId, amount: diff)
        }

        try await expenseRepository.update(expense)
        return expense
    }

    func deleteExpense(id: Int64) async throws {
        let expense = try await expenseRepository.fetch(id: id)
        try await budgetRepository.updateSpentAmount(id: expense.budgetId, amount: -expense.amount)
        try await expenseRepository.delete(id: id)
    }
}
